transmission.go

```diff
@@ -126,78 +127,207 @@ type batch struct {
+	// batch.MarshalJSON likely already did that for us.
+	if len(blob) <= 2 {
 		return
 	}
 
+	// sigh. dislike
 	userAgent := fmt.Sprintf("libhoney-go/%s", version)
 	if UserAgentAddition != "" {
 		userAgent = fmt.Sprintf("%s %s", userAgent, strings.TrimSpace(UserAgentAddition))
 	}
 
-	url := fmt.Sprintf("%s/1/events/%s", e.APIHost, e.Dataset)
-	req, err := http.NewRequest("POST", url, bytes.NewBuffer(blob))
-	req.Header.Set("User-Agent", userAgent)
+	// FML. technically we have nothing that constrains events to a single APIHost
+	// or a single WriteKey per batch. ugggg ben suggestions?
+	var apiHost string
+	var writeKey string
+	for _, events := range b.events {
+		if len(events) < 1 {
+			continue
+		}
+		apiHost = events[0].APIHost
+		writeKey = events[0].WriteKey
+		break
+	}
+
+	reqBody, gzipped := buildReqReader(blob)
+	url := fmt.Sprintf("%s/1/batch", apiHost)
+	req, err := http.NewRequest("POST", url, reqBody)
 	req.Header.Set("Content-Type", "application/json")
-	req.Header.Add("X-Honeycomb-Team", e.WriteKey)
-	req.Header.Add("X-Event-Time", timestamp.Format(time.RFC3339))
-	req.Header.Add("X-Honeycomb-SampleRate", strconv.Itoa(int(e.SampleRate)))
+	if gzipped {
+		req.Header.Set("Content-Encoding", "gzip")
+	}
+	req.Header.Set("User-Agent", userAgent)
+	req.Header.Add("X-Honeycomb-Team", writeKey)
 
+	// send off batch!
 	resp, err := b.httpClient.Do(req)
-
 	end := time.Now().UTC()
 	if b.testNower != nil {
 		end = b.testNower.Now()
 	}
 	dur := end.Sub(start)
-	evResp := Response{}
-	defer func() {
-		if b.blockOnResponses {
-			responses <- evResp
-		} else {
-			select {
-			case responses <- evResp:
-			default:
-				if b.testBlocker != nil {
-					b.testBlocker.Done()
-				}
-			}
-		}
-	}()
-	evResp.Duration = dur
-	evResp.Metadata = e.Metadata
+
+	batchSize := 0
+	for _, events := range b.events {
+		batchSize += len(events) // ... missing out on events dropped during json encoding
+	}
+
 	if err != nil {
-		// TODO add logging or something to raise this error
 		sd.Increment("send_errors")
-		evResp.Err = err
+		// if there's a top-level error, we should send an error response down for
+		// every event?? do we actually have examples that read responses and retry?
+		b.enqueueResponse(Response{
+			Duration: dur,
+			Metadata: fmt.Sprintf("batch containing %d events for %d datasets", batchSize, len(b.events)),
+			Err:      err,
+		})
 		return
 	}
-	sd.Increment("messages_sent")
+
+	sd.Increment("batches_sent")
+	sd.Count("messages_sent", batchSize)
 	defer resp.Body.Close()
-	evResp.StatusCode = resp.StatusCode
-	body, _ := ioutil.ReadAll(resp.Body)
-	evResp.Body = body
+
+	batchResponse := map[string][]Response{}
+	err = json.NewDecoder(resp.Body).Decode(&batchResponse)
+	if err != nil {
+		sd.Increment("response_decode_errors")
+		b.enqueueResponse(Response{
+			Duration: dur,
+			Metadata: fmt.Sprintf("batch containing %d events for %d datasets", batchSize, len(b.events)),
```

Comment:
i think you're right, buuut i think the way we'll do this -> should avoid sending non-event responses